### PR TITLE
Update docu

### PR DESCRIPTION
**Pull request by mbrandenburger**
This PR addresses #124 and #96

### CODE REVIEW

**Comment by mbrandenburger** [comment]
Note that I added a section `## Your first private chaincode` as a placeholder. I suggest that we fill that with content in a separate PR.

File: README.md [hunk 1]
```diff
@@ -230,43 +281,83 @@ enclave [tlcc_enclave](tlcc_enclave) can be build manually.
...
-registry. See [ercc](ercc).
+registry. See [ercc/](ercc).
+
+Moreover, we provide a set of integration tests in [integration/](integration) to demonstrate Fabric Private Chaincode
+capabilities.
+
+
+#### Trouble shooting
+
+Building the project requires docker. We do not recommend to run `sudo make`
```
Comment (g2flyer) [inline README.md]: Maybe backwards pointer to above docker section on how to setup docker to run as user rather than root.?

File: README.md [hunk 2]
```diff
@@ -230,43 +281,83 @@ enclave [tlcc_enclave](tlcc_enclave) can be build manually.
...
 Just follow the description in [fabric](fabric#).
 
+
 # References
 
 - Marcus Brandenburger, Christian Cachin, Rüdiger Kapitza, Alessandro
   Sorniotti: Blockchain and Trusted Computing: Problems, Pitfalls, and a
   Solution for Hyperledger Fabric. https://arxiv.org/abs/1805.08541
 
+
```
Comment (g2flyer) [inline README.md]: Should we mention somewhere the rocketchat channel? Maybe even advertise the Tuesday calls?

File: README.md [hunk 3]
```diff
@@ -230,43 +281,83 @@ enclave [tlcc_enclave](tlcc_enclave) can be build manually.
...
 
 Once you have successfully built the project you can play with the auction chaincode.
 Just follow the description in [fabric](fabric#).
 
+
 # References
 
 - Marcus Brandenburger, Christian Cachin, Rüdiger Kapitza, Alessandro
   Sorniotti: Blockchain and Trusted Computing: Problems, Pitfalls, and a
   Solution for Hyperledger Fabric. https://arxiv.org/abs/1805.08541
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply README.md]: I will add the SRDS paper once I have the DOI.

Yes, We should add the proposal. I will mark it here as WIP.

File: README.md [hunk 4]
```diff
@@ -198,30 +205,74 @@ Place your ias api key and your SPID in the ``ias`` folder as follows:
     echo 'YOUR_SPID' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-private-chaincode/config/ias/spid.txt
 
 
-### Get NanoPB
+###### Working from behind a proxy
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply README.md]: fixed

File: README.md [hunk 5]
```diff
@@ -91,102 +94,106 @@ The system consists of the following components:
...
-## Working from behind a proxy
 
-The current code should work behind a proxy assuming
-  * you have defined the corresponding environment variables (i.e.,
-  `http_proxy`, `https_proxy` and, potentially, `no_proxy`) properly
-  defined, and
-  * docker (daemon & client) is properly set up for proxies as
-    outlined in the Docker documentation for [clients](https://docs.docker.com/network/proxy/) and the [daemon](https://docs.docker.com/config/daemon/systemd/#httphttps-proxy).
-If you run Ubuntu 18.04, make sure you run docker 18.09 or later. Otherwise you will run into problems with DNS resolution inside the container.
+#### Docker
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply README.md]: fixed

File: README.md [hunk 6]
```diff
@@ -198,30 +205,74 @@ Place your ias api key and your SPID in the ``ias`` folder as follows:
...
+insofar that some '0.0.0.0' in ``fabric/sgxconfig/core.yaml`` used by
+clients -- e.g., the peer CLI using the ``address: 0.0.0.0:7051`` config
+as part of the ``peer`` section -- result in the client being unable
+to find the server. The likely error you will see is
+ ``err: rpc error: code = Unavailable desc = transport is closing``.
+In that case, you will have to replace the '0.0.0.0' with a concrete
+ip address such as '127.0.0.1'.
+
+
+##Fabric Private Chaincode
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply README.md]: fixed

File: README.md [hunk 7]
```diff
@@ -198,30 +205,74 @@ Place your ias api key and your SPID in the ``ias`` folder as follows:
...
+
+Clone the code and make sure it is on your `$GOPATH`. (Important: we assume in this documentation and default
+configuration that your `$GOPATH` has a _single_ root-directoy!)
+
+    $ git clone https://github.com/hyperledger-labs/fabric-private-chaincode.git $GOPATH/src/github.com/hyperledger-labs/fabric-private-chaincode
+
+
+#### Patch Fabric
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply README.md]: fixed

File: README.md (hunk 1 above)
Comment (mbrandenburger) [reply README.md]: added

File: README.md (hunk 2 above)
Comment (mbrandenburger) [reply README.md]: this is a great idea. I added a section for that!

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (g2flyer) [reply README.md]: Hmm, i think the proxy section should not be a subsection of the subsection of a subsection but should be at the same level as SGX, docker, Nanobuf and alike?

**Comment by mbrandenburger** [comment]
> One small suggestion (up-level proxy section) but will approve regardless of what you decide to do

I've moved the proxy section now to troubleshooting.

**Comment by sjanakir2015** [comment]
Suggestions: 
1.  a mention of which version of Linux (OS) is supported *before* other dependencies are listed.  https://github.com/intel/linux-sgx lists the supported OS.  
2.  Under Troubleshooting, a summary of expected  environment settings for subsequent (make) commands will be handy from users perspective.  
export SGX_MODE=SIM
export SGX_SSL=/opt/intel/sgxssl
export FABRIC_PATH=${GOPATH}/src/github.com/hyperledger/fabric
export NANOPB_PATH=~/sgx/nanopb/
export SGX_WORK=~/work/src/github.com/hyperledger-labs/fabric-private-chaincode
export NANOPB_PATH=$HOME/sgx/nanopb
export ECC_ATTESTATION_TYPE = epid_unlinkable

From  /opt/intel/sgxsdk/environment:
export SGX_SDK=/opt/intel/sgxsdk
export PATH=$PATH:$SGX_SDK/bin:$SGX_SDK/bin/x64
export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$SGX_SDK/pkgconfig
if [ -z "$LD_LIBRARY_PATH" ]; then
     export LD_LIBRARY_PATH=$SGX_SDK/sdk_libs
else
     export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$SGX_SDK/sdk_libs
fi

3. Under section titled "Building individual components", 
"...  can be build manually" 
to be changed to  "... can be built manually"

**Comment by mbrandenburger** [comment]
> Suggestions:
> 
> 1. a mention of which version of Linux (OS) is supported _before_ other dependencies are listed.  https://github.com/intel/linux-sgx lists the supported OS.
> 2. Under Troubleshooting, a summary of expected  environment settings for subsequent (make) commands will be handy from users perspective.
>    export SGX_MODE=SIM
>    export SGX_SSL=/opt/intel/sgxssl
>    export FABRIC_PATH=${GOPATH}/src/github.com/hyperledger/fabric
>    export NANOPB_PATH=~/sgx/nanopb/
>    export SGX_WORK=~/work/src/github.com/hyperledger-labs/fabric-private-chaincode
>    export NANOPB_PATH=$HOME/sgx/nanopb
>    export ECC_ATTESTATION_TYPE = epid_unlinkable
> 
> From /opt/intel/sgxsdk/environment:
> export SGX_SDK=/opt/intel/sgxsdk
> export PATH=$PATH:$SGX_SDK/bin:$SGX_SDK/bin/x64
> export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$SGX_SDK/pkgconfig
> if [ -z "$LD_LIBRARY_PATH" ]; then
> export LD_LIBRARY_PATH=$SGX_SDK/sdk_libs
> else
> export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$SGX_SDK/sdk_libs
> fi
> 
> 1. Under section titled "Building individual components",
>    "...  can be build manually"
>    to be changed to  "... can be built manually"

Thanks sowmya! I added your suggestions to the docu

**Comment by annumberhocker** [comment]
A few more suggestions: 

- Under build the project:

"Now it's time to build the project. Just hit make and go **grep** a coffee. It may take a few moments when you build the project for the first time."

Need to change grep to grab.

- It is very confusing bouncing back and forth between this README and the fabric/README.md file. This README tells you to go there to patch the fabric, but then that README describes running the auction test case, but you have to bounce back to this README to install and make first, then this README points back to that one to run the auction test.  It would be helpful to have all of the instructions in one place.  I think you could repeat the patch instructions in the **fabric** directory (but have them here as well. they are only a few lines of code to add to this one), but not sure why the instructions to run the auction test case are in that README.  It sets the stage for missing a step or getting confused.

**Comment by mbrandenburger** [comment]
> A few more suggestions:
> 
> * Under build the project:
> 
> "Now it's time to build the project. Just hit make and go **grep** a coffee. It may take a few moments when you build the project for the first time."
> 
> Need to change grep to grab.
> 
> * It is very confusing bouncing back and forth between this README and the fabric/README.md file. This README tells you to go there to patch the fabric, but then that README describes running the auction test case, but you have to bounce back to this README to install and make first, then this README points back to that one to run the auction test.  It would be helpful to have all of the instructions in one place.  I think you could repeat the patch instructions in the **fabric** directory (but have them here as well. they are only a few lines of code to add to this one), but not sure why the instructions to run the auction test case are in that README.  It sets the stage for missing a step or getting confused.

thanks for the comment! I fixed the typo. `fabric/README.md` is now integrated into the main `README.md`

**Comment by annumberhocker** [comment]
Another observation:
If we are running in simulation mode, are we supposed to still have to do the steps in https://github.com/hyperledger-labs/fabric-private-chaincode#intel-attestation-service-ias?

Both Michael and Bruno indicated that my missing config/ias/api_key.txt and spid.txt files should not be an issue if I had SGX_MODE=SIM set.  However, that is not the case.  I have this set and still require these api keys.  So if SIM mode shouldn't require these, this is a bug.  And the README needs to be updated to indicate that https://github.com/hyperledger-labs/fabric-private-chaincode#intel-attestation-service-ias can be optional if SGX_MODE is set to SIM.

**Comment by g2flyer** [comment]
> Another observation:
> If we are running in simulation mode, are we supposed to still have to do the steps in https://github.com/hyperledger-labs/fabric-private-chaincode#intel-attestation-service-ias?

No. But you are referencing the wrong (old) README. At least the new one of this PR (https://github.com/hyperledger-labs/fabric-private-chaincode/blob/f311b32eb49760ce5a3e7bab075f806671a0a17a/README.md) should answer your question in the first paragraph of the IAS section. 

> Both Michael and Bruno indicated that my missing config/ias/api_key.txt and spid.txt files should not be an issue if I had SGX_MODE=SIM set. 

Actually, i didn't say so. Afaik you have to have these files no matter what. I just cross-checked: with the current code you (a) need both files and, while the content of api_key.txt doesn't matter, (b) spid.txt must contain a hex-encoded byte array of length 16.

> However, that is not the case. I have this set and still require these api keys. So if SIM mode shouldn't require these, this is a bug. And the README needs to be updated to indicate that https://github.com/hyperledger-labs/fabric-private-chaincode#intel-attestation-service-ias can be optional if SGX_MODE is set to SIM.

Yep, we should update the doc that even for SIM we require the two files but could give copy/paste instructions like
```
echo 'ignored api-key' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-private-chaincode/config/ias/api_key.txt
echo 'DEADBEEF00000000DEADBEEF00000000' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-private-chaincode/config/ias/spid.txt
```

**Comment by g2flyer** [comment]
btw: given that the doc needs now a revision, we probably also should include a link to the git work flow google doc in the README?

**Comment by mbrandenburger** [comment]
> Yep, we should update the doc that even for SIM we require the two files but could give copy/paste instructions like
> 
> ```
> echo 'ignored api-key' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-private-chaincode/config/ias/api_key.txt
> echo 'DEADBEEF00000000DEADBEEF00000000' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-private-chaincode/config/ias/spid.txt
> ```

I disagree to add this extra in the docu ... if the current code requires these files, we should add them with some dummy content; or change the code that these files are not needed in SIM mode. First, is a good quick fix for this problem.

**Comment by mbrandenburger** [comment]
> btw: given that the doc needs now a revision, we probably also should include a link to the git work flow google doc in the README?

We can do that later as this is not related to the SDKization.

**Comment by g2flyer** [comment]
> > Yep, we should update the doc that even for SIM we require the two files but could give copy/paste instructions like
> > ```
> > echo 'ignored api-key' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-private-chaincode/config/ias/api_key.txt
> > echo 'DEADBEEF00000000DEADBEEF00000000' > ${GOPATH}/src/github.com/hyperledger-labs/fabric-private-chaincode/config/ias/spid.txt
> > ```
> 
> I disagree to add this extra in the docu ... if the current code requires these files, we should add them with some dummy content; or change the code that these files are not needed in SIM mode. First, is a good quick fix for this problem.

hmm, don't think add default files in git is a good idea: it will show up as conflict as soon as edit it (actually, right now it is in .gitignore but leaving it there leads to other problems).
Fixing the code would be the other solution but there is no quick fix there after having looked at it a bit before ..

**Comment by g2flyer** [comment]
> > btw: given that the doc needs now a revision, we probably also should include a link to the git work flow google doc in the README?
> 
> We can do that later as this is not related to the SDKization.

certainly not crucial, but if we think it would be useful to add to contributors, why table it?

**Comment by g2flyer** [comment]
> Fixing the code would be the other solution but there is no quick fix there after having looked at it a bit before ..

Actually, looking again i think there _is_ a quick fix if in `ercc/attestation/ias_credentials/decoration.go` we don't read but put default values if we are in SIM mode

**Comment by g2flyer** [comment]
> > Fixing the code would be the other solution but there is no quick fix there after having looked at it a bit before ..
> 
> Actually, looking again i think there _is_ a quick fix if in `ercc/attestation/ias_credentials/decoration.go` we don't read but put default values if we are in SIM mode

issue 1) and 3) is now addressed in PR #140

**Comment by mbrandenburger** [comment]
> > > btw: given that the doc needs now a revision, we probably also should include a link to the git work flow google doc in the README?
> > 
> > 
> > We can do that later as this is not related to the SDKization.
> 
> certainly not crucial, but if we think it would be useful to add to contributors, why table it given that it is a trival change?

added

**Comment by mbrandenburger** [comment]
> 1. Under https://github.com/mbrandenburger/fabric-secure-chaincode/tree/docu#build-the-project, add:
> 
> ```
> cd $SGX_WORK
> make
> ```

Fixed
 
> Change "successfully patched and build Fabric" to "successfully patched and **built** Fabric"

fixed

> python-protobuf
> clang-format-8 (need to do `sudo mv /usr/bin/clang-format-8 /usr/bin/clang-format` after installed)

Added more details to nanop installation and trouble shooting

**Pull request merged**